index.js

```diff
@@ -70,9 +12,51 @@ function truth() {
  * @param {Array} filter mapbox gl filter
  * @returns {Function} filter-evaluating function
  */
-module.exports = function (filter) {
-    if (!filter) return truth;
-    var filterStr = 'var p = f.properties || f.tags || {}, t = f.type; return ' + compile(filter) + ';';
-    // jshint evil: true
-    return new Function('f', filterStr);
-};
+function createFilter(filter) {
+    return new Function('f', 'return ' + compile(filter));
+}
+
+function compile(filter) {
+    if (!filter || filter.length <= 1) return 'true';
+    var op = filter[0];
+    var str =
+        op === '==' ? compare(filter[1], filter[2], '===', false) :
+        op === '!=' ? compare(filter[1], filter[2], '!==', false) :
+        op === '<' ||
+        op === '>' ||
+        op === '<=' ||
+        op === '>=' ? compare(filter[1], filter[2], op, true) :
+        op === 'any' ? filter.slice(1).map(compile).join('||') :
+        op === 'all' ? filter.slice(1).map(compile).join('&&') :
+        op === 'none' ? '!(' + filter.slice(1).map(compile).join('||') + ')' :
+        op === 'in' ? compileIn(filter[1], filter.slice(2)) :
+        op === '!in' ? '!(' + compileIn(filter[1], filter.slice(2)) + ')' :
+        'true';
+    return '(' + str + ')';
+}
+
+function valueExpr(key) {
+    return key === '$type' ? 'f.type' : '(f.properties || {})[' + JSON.stringify(key) + ']';
```

Comment:
> Does caching affect performance?

As far as I remember, doesn't, but I can readd and make sure now.

> Do we need to keep supporting `f.tags`?

I don't know why they were supported in the first place. When we removed filtering code from gl-js (https://github.com/mapbox/mapbox-gl-js/commit/96013eb9134c50262a61c4a7a4d08d5fa869cddf), it was just `f.properties`, but the first commit here has `f.tags`. Maybe @camilleanne knows?